design-doc.md

```diff
@@ -0,0 +1,128 @@
+
+A per-entry handler of a listener has the means to tell how far "behind" on a stream it is, both index-wise and order-key-wise. This is important for serving guarantees: for example, when a listener is responsible for serving external requests, the common design is to notify its master or load balancer that it is healthy as it is caught up and that it still is in the "replay log" mode.
+
+### Publishing
+
+While publishing, entries should come in increasing order of order keys.
+
+By design, a `Publish()` operation is not thread-safe, and it is advised to acquire stream's publisher within a specific thread that does the publishing.
+
+Implementation-wise, within KnowSheet, this thread would almost always be an HTTP server, a consumer of an in-memory message queue, a listener of another stream or a producer of a TailProduce job. All of the above are single-threaded by design, and they respect the order of calls.
+
+Streams are hardly deleted: in most production systems, if a stream is created it persists for the lifetime of the binary.
+
+With that in mind, however, it is possible to shutdown a stream without terminating the binary. The shutdown operation is designed to be as safe as possible: implementation-wise, it will wait for all pending calls to currently active listeners, as well as signal the acquired publisher to terminate. Only after all the handles have been released, the stream will be destructed.
+
+
+## Replication
+
+From data integrity standpoint, each stream is assigned an **authority**: A binary, running, usually a dedicated machine, that is responsible to reporting the current up-to-date status of the stream.
+
+Sherlock uses master-slave replication. Replication logic is to indefinitely listen to the entries from the index that is the first one not yet published into the local copy of the stream.
+
+(`TODO(dkorolev): Say a few words about our failover philosophy.`)
+
+Data from streams can be replicated. Replication can be thought of as a simple usecase of:
+
+1. A listener factory bound to an HTTP endpoint opening a chunked response HTTP connection ("everybody can 'clone' the stream knowing its name"), and
+2. A publisher, ownership of which is transferred into the HTTP chunked response receiver, so that the only legitimate was to add entries to the stream
+
+(`TODO(dkorolev): Security and authentication.`)
+
+The recommended usecase of Sherlock is to replicate each data stream locally before working with it. This ensures that between re-runs of the binary, only the newly added entries are read by the master transferred over the netwoe
+
+## HEAD Pointer
+
+TBD: A way to notify that the order key has updated without new entries being added, required for real-time stream joining in `TailProduce`.
+
+
+# Implementation Notes
+
+`TODO(dkorolev): Code snippets?`
+
+`TODO(dkorolev): Move all implementation details from above into this section?`
+
```

Comment:
I'd vote for some code snippets in the sections above and all needed elaborations here.
What about a more or less full-fledged example code?